storage/src/checker/mod.rs

```diff
@@ -115,46 +117,57 @@ impl<S: WritableStorage> NodeStore<Committed, S> {
+            Ok(visited) => visited,
+            Err(e) => {
+                return CheckerReport {
+                    errors: vec![e],
+                    high_watermark: db_size,
+                    physical_bytes: 0,
+                    trie_stats: TrieStats::default(),
+                    free_list_stats: FreeListsStats::default(),
+                };
+            }
+        };
 
-        let mut visited = LinearAddressRangeSet::new(db_size)?;
+        let mut errors = Vec::new();
 
         // 2. traverse the trie and check the nodes
         if let Some(progress_bar) = &opt.progress_bar {
             progress_bar.set_length(db_size);
             progress_bar.set_message("Traversing the trie...");
         }
-        let trie_stats = if let (Some(root), Some(root_hash)) =
-            (self.root_as_maybe_persisted_node(), self.root_hash())
-        {
-            // the database is not empty, and has a physical address, so traverse the trie
-            if let Some(root_address) = root.as_linear_address() {
-                self.check_area_aligned(
-                    root_address,
-                    StoredAreaParent::TrieNode(TrieNodeParent::Root),
-                )?;
-                self.visit_trie(
-                    root_address,
-                    root_hash.into_hash_type(),
-                    &mut visited,
-                    opt.progress_bar.as_ref(),
-                    opt.hash_check,
-                )?
-            } else {
-                return Err(CheckerError::UnpersistedRoot);
-            }
-        } else {
-            TrieStats::default()
-        };
+        let trie_stats = self
+            .root_as_maybe_persisted_node()
+            .and_then(|node| self.root_hash().map(|root_hash| (node, root_hash)))
+            .and_then(|(root, root_hash)| {
+                if let Some(root_address) = root.as_linear_address() {
+                    let (trie_stats, trie_errors) = self.visit_trie(
+                        root_address,
+                        root_hash.into_hash_type(),
+                        &mut visited,
+                        opt.progress_bar.as_ref(),
+                        opt.hash_check,
+                    );
+                    errors.extend(trie_errors);
+                    Some(trie_stats)
+                } else {
+                    errors.push(CheckerError::UnpersistedRoot);
+                    None
+                }
+            })
+            .unwrap_or_default();
 
         // 3. check the free list - this can happen in parallel with the trie traversal
         if let Some(progress_bar) = &opt.progress_bar {
             progress_bar.set_message("Traversing free lists...");
         }
-        let free_list_stats = self.visit_freelist(&mut visited, opt.progress_bar.as_ref())?;
+        let (free_list_stats, free_list_traverse_errors) =
+            self.visit_freelist(&mut visited, opt.progress_bar.as_ref());
+        errors.extend(free_list_traverse_errors);
```

Comment:
nit: since this is the only caller, seems like we should just pass `&mut free_list_traverse_errors` to `visit_freelist` instead of allocating a new array and extending errors later.

You might have it this way to support more parallelism, in which case you can ignore this comment.